### PR TITLE
VB-5076 - Create allocation service

### PR DESCRIPTION
**Pull request by adaviesMOJ**
## What does this PR do?
Add new service to handle generating and storing VO / PVOs.

## JIRA Link
https://dsdmoj.atlassian.net/browse/VB-5076

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/service/AllocationService.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    val visitOrders = mutableListOf<VisitOrder>()
+    repeat(prisonIncentiveAmounts.visitOrders) {
+      visitOrders.add(
+        VisitOrder(
+          prisonerId = prisonerId,
+          type = VisitOrderType.VO,
+          status = VisitOrderStatus.AVAILABLE,
+          createdDate = LocalDate.now(),
+          expiryDate = null,
+        ),
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/service/AllocationService.kt:45]: we could use a builder class here.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/service/AllocationService.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+          prisonerId = prisonerId,
+          type = VisitOrderType.PVO,
+          status = VisitOrderStatus.AVAILABLE,
+          createdDate = LocalDate.now(),
+          expiryDate = null,
+        ),
+      )
+    }
+
+    LOG.info("Successfully generated ${visitOrders.size} visit orders for prisonerId $prisonerId")
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/service/AllocationService.kt:60]: should we log VOs and PVOs seperately?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/service/listener/processors/PrisonerConvictionStatusUpdatedProcessor.kt [hunk 3]
```diff
@@ -4,20 +4,25 @@ import com.fasterxml.jackson.databind.ObjectMapper
...
+) {
   companion object {
     val LOG: Logger = LoggerFactory.getLogger(this::class.java)
   }
 
   fun processEvent(domainEvent: DomainEvent) {
     LOG.info("received conviction status changed event: {}", domainEvent)
     val additionalInfo = getAdditionalInfo(domainEvent)
-    // TODO: Call new allocation service - startAllocation method
+    if (additionalInfo.convictedStatus == "Convicted") { // TODO: Confirm how the string will come through
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/service/listener/processors/PrisonerConvictionStatusUpdatedProcessor.kt:23]: enum for ConvictionStatuses?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/config/WebClientConfiguration.kt [hunk 4]
```diff
@@ -1,17 +1,83 @@
...
+  }
+
+  @Bean
+  fun hmppsAuthHealthWebClient(builder: WebClient.Builder): WebClient {
+    return builder.healthWebClient(hmppsAuthBaseUri, healthTimeout)
+  }
+
+  @Bean
+  fun prisonerSearchWebClient(authorizedClientManager: OAuth2AuthorizedClientManager): WebClient {
+    val oauth2Client = getOauth2Client(authorizedClientManager, HmppsAuthClientRegistrationId.PRISONER_SEARCH.clientRegistrationId)
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/config/WebClientConfiguration.kt:38]: We could do this in a more concise way now as per the kotlin template - https://github.com/ministryofjustice/hmpps-template-kotlin/blob/main/src/main/kotlin/uk/gov/justice/digital/hmpps/templatepackagename/config/WebClientConfiguration.kt

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/integration/allocations/AllocationServiceTest.kt [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  @Mock
+  private lateinit var incentivesClient: IncentivesClient
+
+  @Mock
+  private lateinit var visitOrderRepository: VisitOrderRepository
+
+  @InjectMocks
+  private lateinit var allocationService: AllocationService
+
+  @Test
```
Comment (dhirajshettymoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/integration/allocations/AllocationServiceTest.kt:38]: nice!! Presuming we will add more tests as more scenarios get clearer

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/service/AllocationService.kt [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  @Transactional
+  fun startAllocation(prisonerId: String) {
+    LOG.info("Entered AllocationService - startAllocation with prisonerId $prisonerId")
+    visitOrderRepository.saveAll(generatePrisonerVoAndPvo(prisonerId))
+  }
+
+  private fun generatePrisonerVoAndPvo(prisonerId: String): List<VisitOrder> {
+    val prisoner = prisonerSearchClient.getPrisonerById(prisonerId)
+    val prisonerIncentive = incentivesClient.getPrisonerIncentiveReviewHistory(prisoner.prisonerNumber)
+    val prisonIncentiveAmounts = incentivesClient.getPrisonIncentiveLevels(prisoner.prisonId!!, prisonerIncentive.iepCode)
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/service/AllocationService.kt:34]: we could have this as a parameter to the method and if null go to incentivesClient. We can then do the same method call as part of the batch job too.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/integration/allocations/AllocationServiceTest.kt (hunk 5 above)
Comment (adaviesMOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/integration/allocations/AllocationServiceTest.kt:38]: Correct. I believe the running allocation ticket will contain a lot of different edge cases

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/service/AllocationService.kt (hunk 6 above)
Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitallocationapi/service/AllocationService.kt:34]: Yes this is what we discussed on Friday but I believe it will be needed as part of the "running allocation" ticket. I'll leave it as is for this feature, and then look to re-order when we create the cron and other code around running allocations.

**Review by dhirajshettymoj — APPROVED** [review]
LGTM
